Auto refresh API connection

leoccyao opened the pull request:
After interrupted network connectivity (or any other plugin error condition), displayed time entries/reports fail to update until manually reconnecting to the Toggl API. While #149 adds a command to do so without restarting the plugin/Obsidian, I've found it more convenient to trigger an auto refresh on a specified interval.

Adds a setting `autoRefreshInterval` to automatically refresh the Toggl API connection on a set interval. Auto-refresh is disabled by default. As the active timer polling interval is itself 6 seconds, it might be worth changing the unit specified in this setting to be measured in minutes, as opposed to seconds.

mcndt commented on lib/config/DefaultSettings.ts:
```diff
@@ -2,6 +2,7 @@ import type { PluginSettings } from "./PluginSettings";
 
 export const DEFAULT_SETTINGS: PluginSettings = {
   apiToken: null,
+  autoRefreshInterval: 0,
```
Other than pre-fetching the workspace data, the api refresh isn't too heavy. Since this fixes a common problem with users, I would consider turning this one on by default actually. Perhaps at an interval of 60 seconds?

leoccyao replied:
Switched to 1 minute default (and now measured in units of minutes). It's a bit odd to have the refresh interval configurable but not the polling interval; perhaps might be good to also expose the polling interval as a setting, or if changing the interval isn't necessary, to make auto-refresh a simple boolean toggle.

mcndt replied:
I agree with that, I don't like exposing settings to users that could have a one-size-fits-all default.

Since the refresh still fetches some workspace data, a default of 2.5 minutes (150 seconds) seems good to me?

mcndt commented on main.ts:
```diff
@@ -28,6 +28,7 @@ export default class MyPlugin extends Plugin {
     this.toggl = new TogglService(this);
     if (this.settings.apiToken != null || this.settings.apiToken != "") {
       this.toggl.refreshApiConnection(this.settings.apiToken);
+      this.toggl.setRefreshInterval(this.settings.autoRefreshInterval);
```
Since `this.toggl.refreshApiConnection` is async, I think it would make sense to `await` the successful initial setup of the API connection before setting the refresh interval.

Perhaps setting the refresh interval can be done at the end of `refreshApiConnection` instead? Then it's covered by every current (and future) code path that refreshes the connection or changes the API token. 

Currently it's only set on plugin load and encloses the API token set at that time as the token used to refresh.

leoccyao replied:
Hmm, an `await` here also blocks the command registration and other view initializations below it.

Merged the code paths, but now `refreshApiConnection` sets an interval just to destroy it one cycle later. Should work, but doesn't feel like the best way to do it.

mcndt reviewed (changes requested):
Thx for another PR, I'll postpone release of 0.12.0 for this one. 

Can you also resolve the merge conflicts with the main branch?

leoccyao commented:
One thing I noticed---after the refactor to `refreshApiConnection`, the "Reconnecting to Toggl..." message comes up every minute, which seems to be an unintended side effect.

mcndt commented on lib/toggl/TogglService.ts:
```diff
@@ -135,6 +141,13 @@ export default class TogglService {
     }
     apiStatusStore.set(this._ApiAvailable);
+
+    if (interval > 0) {
```
Re: your comment:

> Merged the code paths, but now refreshApiConnection sets an interval just to destroy it one cycle later. Should work, but doesn't feel like the best way to do it.

You're right, the right primitive to use here might be `useTimeout`. Since there's technically no deadline for the `refreshApiConnection` promise to resolve, an interval might start another refresh promise while the previous is still resolving. Setting a timeout for the next refresh at the end of the current one guarantees this cannot happen.

leoccyao replied:
Just to clarify, do you mean `setTimeout`, or a different library?
I'm also not sure if `setTimeout` will be harder to lifecycle manage (whether the loop will automatically get cleaned up when unloading the plugin).

mcndt replied:
`setTimeout` is indeed what I meant. I figured that the Obsidian API would have a similar primitive for cleaning up those like has for `setInterval`, but apparently not. A pity.

What solution would you propose to work with?

leoccyao replied:
New behavior I'm noticing---now when there's a connection error when refreshing, the refresh function errors before starting a new interval. Refreshing then stops until manually refreshing, at which it starts again. Not the behavior we want!

Perhaps refresh logic should check for the existence of an auto-refresh interval, and create if it doesn't exist?

leoccyao commented:
> One thing I noticed---after the refactor to `refreshApiConnection`, the "Reconnecting to Toggl..." message comes up every minute, which seems to be an unintended side effect.

Resolved this issue, updated refreshing logic, and switched to a toggle setting. Hoping this can get merged and a new version released soon!

leoccyao commented:
Has been using these latest changes personally and auto-refresh seems to be working well. Things will get busier for me soon, so let me know if any more changes are needed!

adlnk commented:
@mcndt, any chance of completing this merge? Would love to see this as part of 0.12

LeCheenaX commented:
Waiting for the merge of this PR : ) , for 5 months already though.

LeCheenaX commented:
Considering the project has been abandoned, I implemented [my own plugin](https://github.com/LeCheenaX/WordFlow-Tracker) which completely based on local notes, no Internet needed, and has more capabilities: record changed words, total words, etc.  
![image](https://github.com/user-attachments/assets/996d0e7d-6e62-496d-af1e-959cdfae455b)
